### PR TITLE
QAT Gen4

### PR DESCRIPTION
**Pull request by mythi**
Closes: #451

### CODE REVIEW

File: cmd/qat_plugin/dpdkdrv/dpdkdrv.go [hunk 1]
```diff
@@ -245,56 +245,108 @@ func (dp *DevicePlugin) PostAllocate(response *pluginapi.AllocateResponse) error
...
+		return
+	}
+
+	// No PF devices found, running in a VM?
+	for _, vfDriver := range append([]string{dp.dpdkDriver}, dp.kernelVfDrivers...) {
+		pattern := filepath.Join(dp.pciDriverDir, vfDriver, "????:??:??.?")
+		qatVfDevices = append(qatVfDevices, getPciDevicesWithPattern(pattern)...)
+	}
+
+	return
```
Comment (rojkov) [inline cmd/qat_plugin/dpdkdrv/dpdkdrv.go:296]: By this line my tiny brain forgets what the return value is and the func signature goes out of my terminal viewport.  Can we return `qatVfDevices` explicitly here?

File: cmd/qat_plugin/dpdkdrv/dpdkdrv.go [hunk 2]
```diff
@@ -245,56 +245,108 @@ func (dp *DevicePlugin) PostAllocate(response *pluginapi.AllocateResponse) error
...
+		pattern := filepath.Join(qatPfDevice, "virtfn*")
+		qatVfDevices = append(qatVfDevices, getPciDevicesWithPattern(pattern)...)
+	}
+
+	if len(qatPfDevices) > 0 && len(qatVfDevices) > 0 {
+		return
+	}
+
+	// No PF devices found, running in a VM?
+	for _, vfDriver := range append([]string{dp.dpdkDriver}, dp.kernelVfDrivers...) {
```
Comment (rojkov) [inline cmd/qat_plugin/dpdkdrv/dpdkdrv.go:291]: Is the order of the drivers important here? I guess
```go
for _, vfDriver := range append(dp.kernelVfDrivers, dp.dpdkDriver) {
```
would be more readable.

Comment (mythi) [reply cmd/qat_plugin/dpdkdrv/dpdkdrv.go:291]: We prefer devices that are already bound to the targeted driver. See: #146

File: cmd/qat_plugin/dpdkdrv/dpdkdrv.go (hunk 1 above)
Comment (mythi) [reply cmd/qat_plugin/dpdkdrv/dpdkdrv.go]: makes sense, fixed

File: cmd/qat_plugin/dpdkdrv/dpdkdrv.go [hunk 3]
```diff
@@ -245,56 +245,108 @@ func (dp *DevicePlugin) PostAllocate(response *pluginapi.AllocateResponse) error
...
-	for _, driver := range append([]string{dp.dpdkDriver}, dp.kernelVfDrivers...) {
-		files, err := ioutil.ReadDir(filepath.Join(dp.pciDriverDir, driver))
+
+	for _, vfDevice := range dp.getVfDevices() {
+		vfBdf := filepath.Base(vfDevice)
+
+		vfdevID, err := dp.getDeviceID(vfBdf)
 		if err != nil {
-			klog.Warningf("Can't read sysfs for driver as Driver %s is not available: Skipping", driver)
+			return nil, errors.Wrapf(err, "Cannot obtain device ID for %s", vfBdf)
```
Comment (rojkov) [inline cmd/qat_plugin/dpdkdrv/dpdkdrv.go:318]: nit: no need to embed one more stack trace into the error. Can be fixed in a followup PR though as the existing code in the master branch does the same.

**Review by rojkov — APPROVED** [review]

**Pull request merged**